File: cpp/src/specialized.cpp
#include "tide/specialized.h"
#include "tide/ops.h"
#include "tide/kernel.h"
#include "tide/delivery.h"
#include "stream_support.h"
#include <algorithm>
#include <cmath>
#include <limits>
#include <stdexcept>

namespace tide {
Specialized::Specialized(Graph g, Model m, Options options, std::string topology)
    : graph_(std::move(g)), model_(std::move(m)), options_(options), topology_(std::move(topology)), pool_(options.workers) {
  if (options.profile) throw std::invalid_argument("phase profiling requires Streaming");
  if (options.defer_state_release && !options.compact_events)
    throw std::invalid_argument("deferred state release requires compact events");
  if ((options.packed_sources || options.batch_next) && !options.packed)
    throw std::invalid_argument("packed transport requires packed execution");
  graph_.compile(); configure_model(graph_, model_); validate_model(graph_, model_);
  const Index n = graph_.nodes.size();
  auto require = [](bool ok) { if (!ok) throw std::invalid_argument("specialization topology/options mismatch"); };
  require(topology_ == "self_loop" || topology_ == "ring" || topology_ == "chain" || topology_ == "diamond");
  require(graph_.inputs == std::vector<Index>{0} && graph_.outputs == std::vector<Index>{n - 1});
  std::vector<std::pair<Index, Index>> expected;
  if (topology_ == "diamond") {
    require(n == 4);
    const auto& nodes = graph_.nodes;
    const bool paired = nodes[1].region == nodes[2].region;
    require(nodes[0].region == 0 && nodes[1].region == 1 && nodes[2].region == (paired ? 1 : 2) &&
            nodes[3].region == (paired ? 2 : 3) && graph_.regions.size() == (paired ? 3 : 4));
    expected = {{0, 1}, {0, 2}, {1, 3}, {2, 3}};
    layers_ = paired ? std::vector<std::vector<Index>>{{0}, {1, 2}, {3}}
                     : std::vector<std::vector<Index>>{{0}, {1}, {2}, {3}};
  } else {
    require(static_cast<Index>(graph_.regions.size()) == n);
    for (Index v = 0; v < n; ++v) {
      require(graph_.nodes[v].region == v); layers_.push_back({v});
    }
    const bool cyclic = topology_ == "self_loop" || topology_ == "ring";
    if (topology_ == "self_loop") require(n == 1);
    if (topology_ == "ring") require(n >= 2);
    for (Index v = 0; v < (cyclic ? n : n - 1); ++v) expected.push_back({v, (v + 1) % n});
  }
  require(graph_.edges.size() == expected.size());
  for (size_t i = 0; i < expected.size(); ++i)
    require(std::make_pair(graph_.edges[i].source, graph_.edges[i].target) == expected[i]);
  require(options.mode == "hard" || options.mode == "hst" || options.mode == "softp");
  require(std::isfinite(options.zeta));
}
Result Specialized::run(const Continuation& initial, const std::vector<External>& xs, Index stop, Index seal) {
  Result result; auto& q = result.continuation; q = initial;
  auto atoms = validate_window(graph_, model_, q, xs, stop, seal);
  atoms.insert(atoms.end(), q.pending.begin(), q.pending.end());
  Fibers inbox;
  for (const auto& a : atoms) inbox[{a.batch, a.node, a.time}].push_back(a);
  const Index n = graph_.nodes.size();
  auto execute = [&](const std::vector<Frame>& frames) {
    // Local region-block formulas are shared; the fixed topology owns scheduling
    // and propagation and never invokes Streaming, Frontier or their planner.
    auto events = evaluate_block(graph_, model_, q, frames, inbox, options_, pool_, result.stats);
    for (auto& e : events) {
      deliver(graph_, model_, e, [&](const Atom& a) {
        ++result.stats["visited_edges"];
        inbox[{a.batch, a.node, a.time}].push_back(a);
        if (options_.trace) result.messages.push_back(a);
      }, [&](const Output& output) { result.outputs.push_back(output); });
      ++result.stats["candidate_events"];
      if (options_.trace) result.trace.push_back(std::move(e));
    }
    if (options_.compact_events && !options_.trace) release_stream_events(events, pool_, options_.workers);
  };
  if (topology_ == "self_loop" || topology_ == "ring") {
    for (Index time = q.cut; time < stop; ++time) {
      for (Index node = 0; node < n; ++node) {
        std::vector<Frame> frames;
        for (Index b = 0; b < q.batch_size; ++b)
          if (inbox.count({b, node, time})) frames.push_back({b, node, time, {node}, {}});
        execute(frames);
      }
    }
  } else {
    for (const auto& layer : layers_) {
      // All predecessors of this fixed layer have published their messages.
      // Group actual fibers by region/time, including paired diamond candidates.
      std::map<std::pair<Index, Index>, std::set<Index>> coordinates;
      std::vector<Frame> frames;
      for (const auto& [coordinate, bucket] : inbox) {
        auto [b, v, t] = coordinate;
        if (std::find(layer.begin(), layer.end(), v) != layer.end() && t >= q.cut && t < stop)
          coordinates[{t, b}].insert(v);
      }
      for (const auto& [coordinate, nodes] : coordinates)
        frames.push_back({coordinate.second, graph_.nodes[layer[0]].region, coordinate.first, nodes, {}});
      execute(frames);
    }
  }
  q.pending.clear(); q.cut = stop;
  for (const auto& [coordinate, bucket] : inbox)
    for (const auto& a : bucket) if (a.kind == 1 && a.time >= stop) q.pending.push_back(a);
  std::sort(q.pending.begin(), q.pending.end(), [](const auto& a, const auto& b) { return a.key() < b.key(); });
  std::sort(result.trace.begin(), result.trace.end(), [](const auto& a, const auto& b) {
    return std::tie(a.time, a.batch, a.node) < std::tie(b.time, b.batch, b.node);
  });
  std::sort(result.messages.begin(), result.messages.end(), [&](const auto& a, const auto& b) {
    return std::tie(a.position, a.batch, graph_.edges[a.source].source, a.source)
         < std::tie(b.position, b.batch, graph_.edges[b.source].source, b.source);
  });
  std::sort(result.outputs.begin(), result.outputs.end(), [](const auto& a, const auto& b) {
    return std::tie(a.time, a.batch) < std::tie(b.time, b.batch);
  });
  return result;
}
}  // namespace tide

File: python/tidegraph/specialized.py
"""Independent fixed ring/chain/diamond and layered schedules."""
from collections import defaultdict
import math
from .blocks import canonicalize
from .records import Atom, Result
from .validation import validate_window
from .specialized_step import frame


def validate_topology(graph, topology):
    n = len(graph.nodes)
    if graph.inputs != (0,) or graph.outputs != (n - 1,):
        raise ValueError("specialization requires one endpoint input/output")
    if topology not in {"self_loop", "ring", "chain", "diamond"}:
        raise ValueError("invalid specialization")
    if topology == "diamond":
        regions = tuple(node.region for node in graph.nodes)
        if n != 4 or regions not in {(0, 1, 2, 3), (0, 1, 1, 2)}:
            raise ValueError("diamond requires ordered singleton or paired middle regions")
        expected = [(0, 1), (0, 2), (1, 3), (2, 3)]
        layers = [(0,), (1, 2), (3,)] if regions[1] == regions[2] else [(v,) for v in range(n)]
    else:
        if len(graph.regions) != n or any(node.region != i for i, node in enumerate(graph.nodes)):
            raise ValueError("specialization requires singleton ordered regions")
        if (topology == "self_loop" and n != 1) or (topology == "ring" and n < 2):
            raise ValueError("invalid specialization node count")
        expected = ([(i, (i + 1) % n) for i in range(n)] if topology in {"self_loop", "ring"}
                    else [(i, i + 1) for i in range(n - 1)])
        layers = [(v,) for v in range(n)]
    if [(e.source, e.target) for e in graph.edges] != expected:
        raise ValueError("specialization topology mismatch")
    return layers


def _execute(graph, model, initial, external, stop, seal, layers, cyclic, mode, zeta):
    if mode not in {"hard", "hst", "softp"} or not math.isfinite(zeta):
        raise ValueError("invalid specialization options")
    atoms, ledger = validate_window(graph, model, initial, external, stop, seal)
    q = initial.fork(); q.ledger = ledger
    inbox = defaultdict(list)
    for a in list(q.pending) + atoms:
        inbox[a.node, a.batch, a.time].append(a)
    events, messages, outputs = [], [], []
    def apply(nodes, batch, time):
        block = frame(graph, model, q, nodes, batch, time, inbox, mode, zeta)
        for e in block:
            if not e["active"]:
                continue
            node = e["node"]
            ports = graph.port_indexes[1]
            for slot, payload in e["emitted"].items():
                kind, source = ports.bindings[ports.offsets[node] + slot]
                if kind == 0:
                    outputs.append((batch, time, source, payload * model.output_scale[source]))
                    continue
                edge = graph.edges[source]
                arrival = time + edge.delay
                if arrival >= 2**63:
                    raise ValueError("logical time overflow")
                a = Atom(batch, edge.target, arrival, 1, source, time, payload * model.edge_scale[source])
                inbox[edge.target, batch, arrival].append(a); messages.append(a)
        events.extend(block)
    if cyclic:
        # Strict positive delay makes nodes independent at a fixed logical tick.
        for time in range(q.cut, stop):
            for layer in layers:
                for batch in range(q.batch_size):
                    apply(layer, batch, time)
    else:
        # Fixed dependency layers, not the generic planner or streaming queue.
        for layer in layers:
            coordinates = sorted({(t, b) for (v, b, t), xs in inbox.items()
                                  if v in layer and xs and q.cut <= t < stop})
            for time, batch in coordinates:
                apply(layer, batch, time)
    q.pending = [a for bucket in inbox.values() for a in bucket if a.kind == 1 and a.time >= stop]
    q.cut = stop
    return canonicalize(graph, Result(q, events, outputs, messages, {"candidate_events": len(events)}))


def run(graph, model, initial, external, stop, *, sealed_until, topology, mode="hard", zeta=1.0):
    layers = validate_topology(graph, topology)
    return _execute(graph, model, initial, external, stop, sealed_until, layers,
                    topology in {"self_loop", "ring"}, mode, zeta)


def settle_chain(spec, model, initial, values, *, mode="hard", zeta=1.0):
    """Fixed-chain SettleGraph anchor; independent token propagation schedule."""
    validate_topology(spec.graph, "chain")
    if initial.cut % spec.stride or initial.pending:
        raise ValueError("SettleGraph needs a complete position cut")
    result = run(spec.graph, model, initial, spec.external(values, initial.cut // spec.stride),
                 initial.cut + values.shape[1] * spec.stride,
                 sealed_until=initial.cut + values.shape[1] * spec.stride,
                 topology="chain", mode=mode, zeta=zeta)
    result.outputs = [(b, t - spec.rank(spec.graph.outputs[0]) + spec.output_rank, p, v)
                      for b, t, p, v in result.outputs]
    return result


def settle_layered(spec, model, initial, values, *, mode="hard", zeta=1.0):
    """Fully connected adjacent region layers; independent scalar frame schedule."""
    graph = spec.graph
    layers = [tuple(v for v, node in enumerate(graph.nodes) if node.region == r)
              for r in sorted(range(len(graph.regions)), key=lambda r: spec.ranks[r])]
    expected = sorted((a, b) for left, right in zip(layers, layers[1:]) for a in left for b in right)
    if (sorted((e.source, e.target) for e in graph.edges) != expected or
            sorted(graph.inputs) != list(layers[0]) or sorted(graph.outputs) != list(layers[-1])):
        raise ValueError("layered specialization requires complete adjacent layers and boundaries")
    if initial.cut % spec.stride or initial.pending or values.shape[0] != initial.batch_size:
        raise ValueError("SettleGraph needs a complete position cut and matching batch")
    stop = initial.cut + values.shape[1] * spec.stride
    result = _execute(graph, model, initial, spec.external(values, initial.cut // spec.stride),
                      stop, stop, layers, False, mode, zeta)
    groups = defaultdict(list)
    for b, time, port, value in result.outputs:
        output_time = time - spec.rank(graph.outputs[port]) + spec.output_rank
        groups[b, output_time].append((port, value))
    result.outputs = [(b, t, 0, sum(value for _, value in sorted(pairs)))
                      for (b, t), pairs in sorted(groups.items(), key=lambda item: item[0][::-1])]
    return result
